### PR TITLE
Introduce `MacroDefKind::CustomDerive`

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR introduce a new `MacroDefKind` `CustomDerive`. And use a new `ProcMacroExpander` for its expanding. And the expander is a dummy for now. 

Related: #3654

### CODE REVIEW

File: crates/ra_hir_def/src/nameres/collector.rs [hunk 1]
```diff
@@ -854,6 +878,28 @@ impl ModCollector<'_, '_> {
         }
     }
 
+    fn collect_proc_macro(&mut self, attrs: &Attrs) {
+        if let Some(derive_subtree) = attrs.by_key("proc_macro_derive").tt_values().next() {
```
Comment (matklad) [inline crates/ra_hir_def/src/nameres/collector.rs]: Why do we need to collect proc_macro *definitions* here? 

I think the more correct approach would be to collect the set of procmacros from the compiled proc_macro code.

For example, if we had `foo` proc macro, than run `cargo check` to compile it, and then rename it to `bar`, *without* re-runing `cargo check`, we should resole and epxand `foo` and not `bar`.

Comment (edwin0cheng) [reply crates/ra_hir_def/src/nameres/collector.rs]: Um.. the original idea is, we might want to get back some location information from the proc-macro definition in the feature, but I think your approach is reasonable too.

**Comment by edwin0cheng** [comment]
I move the code to `DefCollector::collect` and change it to merely register an array of `MacroExpander` which passed from `CrateGraph` (Right now it is empty).

File: crates/ra_hir_def/src/nameres/collector.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir_def/src/nameres/collector.rs]: Yeah, I think long-term we need to do both (and, in general, have capabilities for relating compiled code (proc macros and/or rlibs) and source code), but it's important to not mix the two accidently, so I'd skip this for now.

**Pull request merged**